Add Postgres to Mathesar docker image

silentninja opened the pull request:
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #3072
The PR adds a Postgres server to the docker image. The in-built database server starts only if the environment variable `DJANGO_DATABASE_URL` is not set, this is done to reduce unnecessary resources. 


The following behavior takes place when `DJANGO_DATABASE_URL` env variable is not set
- The inbuilt Postgres server is started
- Default DB user is created with 
  -  username  as `mathesar` 
  -  password  as `mathesar` 
- Default database is created with
  -  database name as `mathesar_django`
- The  `DJANGO_DATABASE_URL` env variable is set to point to the default database running inside the docker image
   - The connection string value for the env variable will be `postgres://mathesar:mathesar@localhost:5432/mathesar_django`
- The  value of the  `MATHESAR_DATABASES` env variable is set to `(mathesar_tables|postgresql://mathesar:mathesar@localhost:5432/mathesar)`


**Technical details**
Most of the information is the same as the [official Postgres Dockerfile](https://github.com/docker-library/postgres/blob/43d17d5ced92f230fa8c196e746f2e2aa288e5e8/13/bullseye/Dockerfile). I made the following changes
- Added the Postgres repo for installing Postgres 13. Debian buster(used by our docker image) comes with Postgres 11 by default
- Set the following default credentials for the database that is created initially
  -  username  as `mathesar` 
  -  password  as `mathesar` 
  -  database as `mathesar_django`

  

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

mathemancer commented on Dockerfile:
```diff
@@ -3,7 +3,112 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
+
+# grab gosu for easy step-down from root
+# https://github.com/tianon/gosu/releases
+ENV GOSU_VERSION 1.14
+RUN set -eux; \
+	savedAptMark="$(apt-mark showmanual)"; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends ca-certificates wget; \
+	rm -rf /var/lib/apt/lists/*; \
+	dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"; \
+	wget -O /usr/local/bin/gosu "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch"; \
+	wget -O /usr/local/bin/gosu.asc "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch.asc"; \
+	export GNUPGHOME="$(mktemp -d)"; \
+	gpg --batch --keyserver hkps://keys.openpgp.org --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4; \
+	gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu; \
+	gpgconf --kill all; \
+	rm -rf "$GNUPGHOME" /usr/local/bin/gosu.asc; \
+	apt-mark auto '.*' > /dev/null; \
+	[ -z "$savedAptMark" ] || apt-mark manual $savedAptMark > /dev/null; \
+	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false; \
+	chmod +x /usr/local/bin/gosu; \
+	gosu --version; \
+	gosu nobody true
```
Also the certificates and wget are already installed, I believe. Alsoalso, I don't think we need `wget`. Just use curl.

silentninja replied:
I got a package missing error when I tried to install it from the Debian repo and I ended up copying the instructions from the gosu repo. But it was an oversight on my end, I replaced it with installing from debian repo.

mathemancer commented on Dockerfile:
```diff
@@ -3,7 +3,112 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
+# make the "en_US.UTF-8" locale so postgres will be utf-8 enabled by default
+RUN set -eux; \
+	if [ -f /etc/dpkg/dpkg.cfg.d/docker ]; then \
+# if this file exists, we're likely in "debian:xxx-slim", and locales are thus being excluded so we need to remove that exclusion (since we need locales)
```
We're not, right? We're starting from the python buster image, which isn't on slim.

mathemancer commented on Dockerfile:
```diff
@@ -3,7 +3,112 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
+RUN set -ex; \
+    curl https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - ; \
+    echo "deb http://apt.postgresql.org/pub/repos/apt/ buster-pgdg main" > /etc/apt/sources.list.d/pgdg.list; \
+    mkdir -p /etc/postgresql-common/createcluster.d/; \
+    echo "create_main_cluster = false" > /etc/postgresql-common/createcluster.d/no-main-cluster.conf; \
```
Why specify this? What's wrong with the defaults?

mathemancer commented on Dockerfile:
```diff
@@ -3,7 +3,112 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
+
+ENV PG_VERSION 13.11-1.pgdg110+1
+
+# make the sample config easier to munge (and "correct by default")
+RUN set -eux; \
+	dpkg-divert --add --rename --divert "/usr/share/postgresql/postgresql.conf.sample.dpkg" "/usr/share/postgresql/$PG_MAJOR/postgresql.conf.sample"; \
+	cp -v /usr/share/postgresql/postgresql.conf.sample.dpkg /usr/share/postgresql/postgresql.conf.sample; \
+	ln -sv ../postgresql.conf.sample "/usr/share/postgresql/$PG_MAJOR/"; \
+	sed -ri "s!^#?(listen_addresses)\s*=\s*\S+.*!\1 = '*'!" /usr/share/postgresql/postgresql.conf.sample; \
+	grep -F "listen_addresses = '*'" /usr/share/postgresql/postgresql.conf.sample
+
```
Why does this part exist? The comment doesn't help me understand.

mathemancer commented on Dockerfile:
```diff
@@ -3,7 +3,112 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
+	sed -ri "s!^#?(listen_addresses)\s*=\s*\S+.*!\1 = '*'!" /usr/share/postgresql/postgresql.conf.sample; \
+	grep -F "listen_addresses = '*'" /usr/share/postgresql/postgresql.conf.sample
+
+RUN mkdir -p /var/run/postgresql && chown -R postgres:postgres /var/run/postgresql && chmod 2777 /var/run/postgresql
```
Why?? All these things are set up by installing postgres, right?

mathemancer reviewed (changes requested):
My main overall comment is that this is _waay_ more complicated than it warrants. I can't see any advantages we're getting over just 
- adding the postgres repo
- Installing postgres
- Running it with `pg_ctl`  (and enabling it so it starts whenever you start the container)

We don't need to be able to do 90% of the stuff enabled by the run script, IMO, and using the above would avoid all manner of weird `chmod`ding and `chroot`ing.

I can already tell by looking at this that no one's going to want to maintain it, and we're going to end up with weird cruft, things you cant touch without breaking, etc. And out-of-date versions of packages.

silentninja commented:
Thanks for the review @mathemancer. I removed most of the irrelevant steps. I didn't remove the steps that change the permissions during runtime because we need it for users running docker without root privileges.

kgodey reviewed (approved):
Made some minor updates to docs. @silentninja you forgot to fix the nav and the main page, I updated those too.

I did not review anything except the docs.

@mathemancer please also review the docs.

mathemancer commented on Dockerfile:
```diff
@@ -3,7 +3,49 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
+# flush tables to disk, which is the best compromise available to avoid data
+# corruption.
+#
+# Users who know their applications do not keep open long-lived idle connections
+# may way to use a value of SIGTERM instead, which corresponds to "Smart
+# Shutdown mode" in which any existing sessions are allowed to finish and the
+# server stops when all sessions are terminated.
+#
+# See https://www.postgresql.org/docs/12/server-shutdown.html for more details
+# about available PostgreSQL server shutdown signals.
+#
+# See also https://www.postgresql.org/docs/12/server-start.html for further
+# justification of this as the default value, namely that the example (and
+# shipped) systemd service files use the "Fast Shutdown mode" for service
+# termination.
+#
```
I think you can remove this part of the comment.

mathemancer commented on docs/docs/installation/docker/index.md:
````diff
@@ -19,46 +19,18 @@ You should have permission to run Docker containers on the system.
 
     ```bash
     docker run \
       --detach
````
missing backslash

mathemancer commented on db-run.sh:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
```
I think you should `set -e` here. There's no point in continuing if this breaks (is there?).

mathemancer commented on db-run.sh:
```diff
@@ -0,0 +1,25 @@
+docker_setup_env() {
+	declare -g DATABASE_ALREADY_EXISTS
+	# look specifically for PG_VERSION, as it is expected in the DB dir
+	if [ -s "$PGDATA/PG_VERSION" ]; then
```
This test is defective. 
- Missing a `$`
- `$PG_VERSION` is not set.

Wouldn't the major version suffice anyway?

silentninja replied:
No, the test is correct. `PG_VERSION` is a file inside the data directory which is created when the cluster created.

mathemancer reviewed (changes requested):
This is a major improvement over the last iteration. Thank you for that!

However, while it works in the docker compose setup, it doesn't work when run as a single image. The problem is in the `db-run.sh` script. It has a couple of typos that make the 'is there already a db?' test defective.

### Docs Review
- I suspect we can lower the required docker version by alot. Maybe not worth worrying about right now
- The uninstalling instructions are out-of-sync for the docker install. If they follow the instructions for installing, they won't have the separate DB to worry about.
- There's a missing backslash. See my specific comment.
- We need to add instructions for starting/stopping the service.

silentninja commented:
@mathemancer I ran across a lot of bugs  or unintented effects trying to use `pg_createcluster`. These are the options I tried

#### pg_createcluster (Works with some inconvenience to the user)

When creating a cluster with `pg_createcluster`, it ends up creating the conf directory in `/etc/postgresql/<version>/<cluster-name>/` instead of the data directory which is in `/var/lib/postgresql/<cluster-name>`. The config files are needed for starting the cluster.  So now the user has to mount `/etc/postgresql/<pg-version>/<cluster-name>/` in addition to the data directory, So in case of a postgres version update the user needs to make changes to the mounted volume path.

#### initdb and pg_ctl (Fails due to permission errors)

I tried looking into `initdb` and `pg_ctl` as a alternative. Both of these tools cannot be run as root and docker initializes a volume folder with root privileges(pg_createcluster changes the ownership during the runtime as it is ran as root process). So `initdb` does cannot change the privilege of the data directory(mounted as a volume) and we need to change the privileges before using `initdb` as a postgres user which is what the code in the initial commit of the PR was actually doing.
 
 
 #### Mix of both pg_createcluster and pg_ctl(Fails on restart)
 
I thought I finally got something working with a good UX in this commit but unfortunately it has its own set of problems too. when the postgres server it always starts in recovery mode. @mathemancer Is there anything I am missing here

kgodey commented:
> I suspect we can lower the required docker version by alot. Maybe not worth worrying about right now

Please create an issue to track this.

mathemancer commented on Dockerfile:
```diff
@@ -3,24 +3,24 @@ ARG PYTHON_REQUIREMENTS=requirements.txt
-
-VOLUME /var/lib/postgresql/data
+ENV PGDATA /var/lib/postgresql/mathesar
+VOLUME /etc/postgres/15/mathesar
```
```suggestion
VOLUME /etc/postgresql/15/mathesar
```
Also, why not mount `/etc/postgresql/` instead? I thought you were going to try that so we can avoid fiddling around with mounting different version numbers if we want to upgrade and so on.

mathemancer reviewed (changes requested):
I think this simplifies things nicely, but:
- there are a number of unaddressed comments from my previous review; some are important.
- I think we need to block this on compatibility with PostgreSQL 15. There's (at least one) issue that keeps the tests from passing on PG15.
- See my specific comment about the mount point

silentninja commented:
@mathemancer I made changes to Github action to use the docker image and pass the credentials as an environment variable when calling in docker-exec (to avoid breaking any workflow using docker-compose). So we are using the inbuilt postgres server to test Postgres version 13,14 and 15

silentninja commented:
@mathemancer I addressed all the review feedback, and added the docker image to our test suite. The docs don't need to be updated as there aren't any breaking changes to our test runner. 
I would like to get a final review to make sure your concerns are addressed.

mathemancer reviewed (commented):
Overall, I think this is getting much better! However:
- Running the container twice results in an error:
```
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
```
See my specific  comment about the documented `run` command.

mathemancer commented on docs/docs/installation/docker/index.md:
```diff
@@ -19,49 +19,17 @@ You should have permission to run Docker containers on the system.
       -v static:/code/static \
       -v media:/code/media \
+      -v postgresql_config:/etc/postgresql/ \
+      -v postgresql_data:/var/lib/postgresql/data \
```
Is this correct? Am I missing something? I think this is misaligned with the volume specified in the dockerfile.

mathemancer reviewed (changes requested):
Sorry, mis-clicked on my previous review. It should have been 'request changes' and included the specific comment about the volume mountpoint.

mathemancer reviewed (changes requested):
I think the config volume still doesn't match.

mathemancer reviewed (approved):
Okay, I think this is working, and I can find no other problems.

This is a real improvement to the installation for a user trying out Mathesar, nice work!
